Packageset membership

bryceharrington opened the pull request:
First cut at a page explaining packageset membership.

Yes, there is already official documentation in the main wiki but this combines that with advice from https://wiki.ubuntu.com/RobieBasak/DMB/CoreDev and presents the information in more of a concise worksheet geared to server team members.

The other commit updates some of the examples from bionic to focal, where it could be done with no impact to the example (i.e. by making version numbers inaccurate).

basak commented on MembershipInPackageSet.md:
```diff
@@ -0,0 +1,80 @@
+Membership in Package Set
+-------------------------
+
+Upload rights can be given for certain package-sets, such as 'server' or 'desktop'.  It's original design intent was for individuals who will only be working on a very small set of packages.
```
It's -> Its

basak commented on MembershipInPackageSet.md:
```diff
@@ -0,0 +1,80 @@
+Application Process
+-------------------
+
+0.  (Optional) If you have been involved in the Ubuntu community for a while in any fashion and you're not already a regular Ubuntu Member, add yourself to the next [Ubuntu Membership Board meeting](https://wiki.ubuntu.com/Membership/Boards), following its guidelines for what to include in your wiki page.
```
The general bar is "sustained and significant contributions" rather than "any fashion" (from https://wiki.ubuntu.com/Membership/NewMember). However if the sustained and significant contributions are in Ubuntu development, then [Contributing Developer](https://wiki.ubuntu.com/UbuntuDevelopers#ContribDev) is perhaps more appropriate since the bar is otherwise the same. This would be reviewed by the DMB rather than the general Ubuntu membership boards, and a successful applicant is implicitly granted Ubuntu Membership. So if recommending anything at all, I'd recommend this over seeking Ubuntu Membership for developers - unless this person has made "sustained and significant contributions" in non-developer areas in Ubuntu.

But it also seems like an extra unnecessary step to me to even suggest it, unless a contributor intends to stop there, or wants progressive recognition. I can only recall one occasion where it was used in the past few years, and that was an exceptional case. Other than that, I don't think we've recommended this path before, unless it was before my time (pre 2011).

basak commented on MembershipInPackageSet.md:
```diff
@@ -0,0 +1,80 @@
+  * [Ubuntu's release process](https://wiki.ubuntu.com/UbuntuDevelopment/ReleaseProcess), including the
+    [freeze exception process](https://wiki.ubuntu.com/FreezeExceptionProcess)
+  * Running [Autopkgtest](PackageTests)
+  * Troubleshooting migration of packages from -proposed
```
Link: https://wiki.ubuntu.com/ProposedMigration

basak reviewed (approved):
Great writeup! I added some inline suggestions, but please take them or leave them as you feel appropriate. It's great that you've written this and I want to avoid bikeshedding on what is fundamentally very subjective.

utkarsh2102 commented on MembershipInPackageSet.md:
```diff
@@ -0,0 +1,80 @@
+
+Further Information
+-------------------
+https://wiki.ubuntu.com/RobieBasak/DMB/CoreDev
```
This line needs two spaces in the end for Markdown to render it correctly.
```suggestion
https://wiki.ubuntu.com/RobieBasak/DMB/CoreDev  
```

utkarsh2102 commented on MembershipInPackageSet.md:
```diff
@@ -0,0 +1,80 @@
+
+Once you've been granted packageset upload permissions, there's at least three distinct directions you could embark on, depending on your goals:
+
+  * MOTU, if you want to get depth into the Ubuntu packaging world.
```
Should be ideal to put in a link for those who don't know MOTU (yet).
```suggestion
  * [MOTU](https://wiki.ubuntu.com/MOTU), if you want to get depth into the Ubuntu packaging world.
```

utkarsh2102 commented on MembershipInPackageSet.md:
```diff
@@ -0,0 +1,80 @@
+https://wiki.ubuntu.com/RobieBasak/DMB/CoreDev
+https://wiki.ubuntu.com/UbuntuDevelopers
+
+Once you've been granted packageset upload permissions, there's at least three distinct directions you could embark on, depending on your goals:
```
```suggestion
Once you've been granted packageset upload permissions, there are at least three distinct directions you could embark on, depending on your goals:
```

utkarsh2102 reviewed (approved):
This is great; clear and concise. There are a few minor nitpicks I highlighted below (with the suggestions, so you can commit directly with a click).

Some stuff here and there could be also worked on (proof-reading, et al) but that's really, really minor. And I can do this later when it's merged.

In all, this is awesome, thanks for writing this! 😄

bryceharrington commented:
On Wed, Mar 17, 2021 at 06:49:28PM -0700, basak wrote:
> @basak approved this pull request.
> 
> Great writeup! I added some inline suggestions, but please take them or leave them as you feel appropriate. It's great that you've written this and I want to avoid bikeshedding on what is fundamentally very subjective.
> 

Thanks for reviewing, I've incorporated the suggested changes.

> +Application Process
> +-------------------
> +
> +0.  (Optional) If you have been involved in the Ubuntu community for a while in any fashion and you're not already a regular Ubuntu Member, add yourself to the next [Ubuntu Membership Board meeting](https://wiki.ubuntu.com/Membership/Boards), following its guidelines for what to include in your wiki page.
> 
> The general bar is "sustained and significant contributions" rather than "any fashion" (from https://wiki.ubuntu.com/Membership/NewMember). However if the sustained and significant contributions are in Ubuntu development, then [Contributing Developer](https://wiki.ubuntu.com/UbuntuDevelopers#ContribDev) is perhaps more appropriate since the bar is otherwise the same. This would be reviewed by the DMB rather than the general Ubuntu membership boards, and a successful applicant is implicitly granted Ubuntu Membership. So if recommending anything at all, I'd recommend this over seeking Ubuntu Membership for developers - unless this person has made "sustained and significant contributions" in non-developer areas in Ubuntu.
> 
> But it also seems like an extra unnecessary step to me to even suggest it, unless a contributor intends to stop there, or wants progressive recognition. I can only recall one occasion where it was used in the past few years, and that was an exceptional case. Other than that, I don't think we've recommended this path before, unless it was before my time (pre 2011).

Mulling this over a bit more, I think the thing I like about this step
is that it gives them an immediate task they can accomplish.  The value
here is the initial creation of their wiki page.  For a new employee
this will be easy for them to do while they're already still in "sell
yourself" mode.

Beyond that, you're right this is probably unnecessary.  I'll go ahead
and simplify it down to just setting up the wiki page and dispense with
formally applying for anything.

> > +  * Creating the initial packaging for new software
> +  * Merging updates from Debian
> +  * Backporting patches
> +  * Forwarding bugs and patches to Debian and to upstream maintainers
> +  * Stable Release Update process
> +
> +Make sure you've done each of the above items at least a couple of times.  If you're not comfortable with any of these, plan on doing some additional practice with them.
> +
> +You should also work towards understanding some more advanced packaging topics:
> +
> +  * Purpose of the [different files in debian/](https://packaging.ubuntu.com/html/debian-dir-overview.html)
> +  * [Debian policy](http://www.debian.org/doc/debian-policy/)
> +  * [Ubuntu's release process](https://wiki.ubuntu.com/UbuntuDevelopment/ReleaseProcess), including the
> +    [freeze exception process](https://wiki.ubuntu.com/FreezeExceptionProcess)
> +  * Running [Autopkgtest](PackageTests)
> +  * Troubleshooting migration of packages from -proposed
> 
> Link: https://wiki.ubuntu.com/ProposedMigration

Thanks,
Bryce

bryceharrington commented:
On Thu, Mar 18, 2021 at 07:27:09AM -0700, Utkarsh Gupta wrote:
> @utkarsh2102 approved this pull request.
> 
> This is great; clear and concise. There are a few minor nitpicks I highlighted below (with the suggestions, so you can commit directly with a click).
> 
> Some stuff here and there could be also worked on (proof-reading, et al) but that's really, really minor. And I can do this later when it's merged.
> 
> In all, this is awesome, thanks for writing this! 😄 
> 
> > +While you may not have direct experience with some or most of these topics, you should at least be conversant in all of them conceptually.
> +
> +In addition to Ubuntu packaging, you will need to have some technical expertise with the subset of software you'll be working on, and the processes and procedures standardized for them.  For example, for the Ubuntu server team you would need to have experience with technologies such as systemd and networking, and processes such as using git-ubuntu for package maintenance.
> +
> +Keep in mind the DMB's perspective will follow the "Need to unblock" principle:  They want to approve applications that will help either reduce contributor friction or to save work of sponsors.  Establishing yourself as an area expert that is a resource for contributors helps prove the former, and your volume and frequency of uploads justifies the latter.
> +
> +Collaboration with the upstream(s) related to the software within your packageset is important as well.  Gain experience with your upstreams' bug reporting processes, patch review processes, and testing/CI systems if you haven't so far.  In general, the narrower the number of packages you plan to focus on, the stronger your collaboration with upstream you should be able to show.
> +
> +And on the flip side, if you are a Canonical employee you will need to be able to negotiate the distinction between Ubuntu governance and Canonical priorities, within your area of focus.  Sometimes these can appear to collide and require a certain level of diplomacy to find solutions that work well for both sides.  The less experience you have in Debian or other open source communities, the more thought you'll want to put into this.
> +
> +Finally, you'll know you're past ready for applying if anyone ever asks, "How do you not already have upload rights??"
> +
> +
> +Further Information
> +-------------------
> +https://wiki.ubuntu.com/RobieBasak/DMB/CoreDev
> 
> This line needs two spaces in the end for Markdown to render it correctly.
> ```suggestion
> https://wiki.ubuntu.com/RobieBasak/DMB/CoreDev  
> ```

Yeah, I'll just go ahead and make them bullet lists.

> +Collaboration with the upstream(s) related to the software within your packageset is important as well.  Gain experience with your upstreams' bug reporting processes, patch review processes, and testing/CI systems if you haven't so far.  In general, the narrower the number of packages you plan to focus on, the stronger your collaboration with upstream you should be able to show.
> +
> +And on the flip side, if you are a Canonical employee you will need to be able to negotiate the distinction between Ubuntu governance and Canonical priorities, within your area of focus.  Sometimes these can appear to collide and require a certain level of diplomacy to find solutions that work well for both sides.  The less experience you have in Debian or other open source communities, the more thought you'll want to put into this.
> +
> +Finally, you'll know you're past ready for applying if anyone ever asks, "How do you not already have upload rights??"
> +
> +
> +Further Information
> +-------------------
> +https://wiki.ubuntu.com/RobieBasak/DMB/CoreDev
> +https://wiki.ubuntu.com/UbuntuDevelopers
> +
> +Once you've been granted packageset upload permissions, there's at least three distinct directions you could embark on, depending on your goals:
> +
> +  * MOTU, if you want to get depth into the Ubuntu packaging world.
> 
> Should be ideal to put in a link for those who don't know MOTU (yet).
> ```suggestion
>   * [MOTU](https://wiki.ubuntu.com/MOTU), if you want to get depth into the Ubuntu packaging world.
> ```

Actually, I have a followup page for that, which I'll link to from here.
I was originally going to submit them together, but figured I should get
this page in asap since the topic was pertinent.

> +Keep in mind the DMB's perspective will follow the "Need to unblock" principle:  They want to approve applications that will help either reduce contributor friction or to save work of sponsors.  Establishing yourself as an area expert that is a resource for contributors helps prove the former, and your volume and frequency of uploads justifies the latter.
> +
> +Collaboration with the upstream(s) related to the software within your packageset is important as well.  Gain experience with your upstreams' bug reporting processes, patch review processes, and testing/CI systems if you haven't so far.  In general, the narrower the number of packages you plan to focus on, the stronger your collaboration with upstream you should be able to show.
> +
> +And on the flip side, if you are a Canonical employee you will need to be able to negotiate the distinction between Ubuntu governance and Canonical priorities, within your area of focus.  Sometimes these can appear to collide and require a certain level of diplomacy to find solutions that work well for both sides.  The less experience you have in Debian or other open source communities, the more thought you'll want to put into this.
> +
> +Finally, you'll know you're past ready for applying if anyone ever asks, "How do you not already have upload rights??"
> +
> +
> +Further Information
> +-------------------
> +https://wiki.ubuntu.com/RobieBasak/DMB/CoreDev
> +https://wiki.ubuntu.com/UbuntuDevelopers
> +
> +Once you've been granted packageset upload permissions, there's at least three distinct directions you could embark on, depending on your goals:
> 
> ```suggestion
> Once you've been granted packageset upload permissions, there are at least three distinct directions you could embark on, depending on your goals:
> ```
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/canonical/ubuntu-maintainers-handbook/pull/11#pullrequestreview-615446004

Thanks, I've squashed these changes in, and force pushed.

I'm going to add a bit more directions for filling out the application
itself and will go ahead and land it.

Bryce

cpaelzer commented on Concepts.md:
```diff
@@ -140,17 +140,27 @@ Apt uses `sources.list` and `sources.list.d` to tell it which suites to use.
-    deb-src http://de.archive.ubuntu.com/ubuntu/ bionic main restricted universe
+    deb http://archive.ubuntu.com/ubuntu/ focal main restricted universe
+    deb-src http://archive.ubuntu.com/ubuntu/ focal main restricted universe
 
```
Yeah that might have been examples from me or from when Karl was in Germany.
Thanks for clearing this out.

cpaelzer commented on MembershipInPackageSet.md:
```diff
@@ -0,0 +1,81 @@
+  * [Ubuntu's release process](https://wiki.ubuntu.com/UbuntuDevelopment/ReleaseProcess), including the
+    [freeze exception process](https://wiki.ubuntu.com/FreezeExceptionProcess)
+  * Running [Autopkgtest](PackageTests)
+  * Troubleshooting [https://wiki.ubuntu.com/ProposedMigration](migration of packages) from -proposed
```
IMHO this (the "you don't need to be an expert, but understand the concepts") list needs to have a few more added:
- library transitions https://people.canonical.com/~ubuntu-archive/transitions/
- Seed management https://wiki.ubuntu.com/SeedManagement & https://people.canonical.com/~ubuntu-archive/component-mismatches-proposed.html
- NBS https://people.canonical.com/~ubuntu-archive/nbs.html
- Multiarch https://wiki.ubuntu.com/MultiarchCross

Those are in order of importance and e.g. the top two came up in my CoreDev application.

cpaelzer reviewed (commented):
A very nice summary for our coming new team members indeed - thanks!
I'm mostly ok with all of that and couldn't have done better, especially after Robie and Utkarsh already reviewed there isn't much left.
The one thing I'd change before merging this is in the list of "topcis to be aware of" (see my inline comment there).

cpaelzer commented on PackageTests.md:
```diff
@@ -60,7 +60,7 @@ Everything after the `--` tells it how to run the tests. `qemu` is shorthand for
 
-    $ autopkgtest -U -s -o dep8-mypackage-ppa --setup-commands="sudo add-apt-repository -y -u -s ppa:mylaunchpaduser/bionic-mypackage-fixed-something-1234567" -B mypackage -- qemu /var/lib/adt-images/autopkgtest-bionic-amd64.img
+    $ autopkgtest -U -s -o dep8-mypackage-ppa --setup-commands="sudo add-apt-repository -y -u -s ppa:mylaunchpaduser/focal-mypackage-fixed-something-1234567" -B mypackage -- qemu /var/lib/adt-images/autopkgtest-focal-amd64.img
 
```
As it is a common issue once you need cross-arch, what do you think about also adding a small
"how to run in canonistack" section based on and linking to 
https://wiki.canonical.com/InformationInfrastructure/IS/CanoniStack-BOS01 and https://wiki.ubuntu.com/ProposedMigration#Reproducing_tests_in_the_cloud ?

basak commented:
On Thu, Mar 18, 2021 at 11:49:48PM -0700, Christian Ehrhardt wrote:
> IMHO this (the "you don't need to be an expert, but understand the concepts") list needs to have a few more added:
> - library transitions https://people.canonical.com/~ubuntu-archive/transitions/
> - Seed management https://wiki.ubuntu.com/SeedManagement & https://people.canonical.com/~ubuntu-archive/component-mismatches-proposed.html
> - NBS https://people.canonical.com/~ubuntu-archive/nbs.html
> - Multiarch https://wiki.ubuntu.com/MultiarchCross

These are core dev specific though, and this is a packageset-specific page?

bryceharrington commented:
On Thu, Mar 18, 2021 at 11:49:48PM -0700, Christian Ehrhardt wrote:
> IMHO this (the "you don't need to be an expert, but understand the concepts") list needs to have a few more added:
> - library transitions https://people.canonical.com/~ubuntu-archive/transitions/
> - Seed management https://wiki.ubuntu.com/SeedManagement & https://people.canonical.com/~ubuntu-archive/component-mismatches-proposed.html
> - NBS https://people.canonical.com/~ubuntu-archive/nbs.html
> - Multiarch https://wiki.ubuntu.com/MultiarchCross
> 
> Those are in order of importance and e.g. the top two came up in my CoreDev application.

Yep, these are good topics, and I have most of them in the corresponding
MOTU doc.  That one goes into more depth also about establishing trust.
I hope to get a PR for the MOTU doc posted soon.

But for the *packageset* directions I think it is more helpful if we're
selective in what we suggest, since we could easily overwhelm someone
that was on the newer side, and let the focus be stronger on basics.  So
that's why some of the archive management topics are conspicuously
absent here.

(There's also a *coredev* doc, but it's training section is basically
just "gain mastery through practical experience in all of the above
topics".)

Thanks,
Bryce

bryceharrington commented:
On Thu, Mar 18, 2021 at 11:51:26PM -0700, Christian Ehrhardt wrote:
> @cpaelzer commented on this pull request.
> 
> 
> 
> > @@ -60,7 +60,7 @@ Everything after the `--` tells it how to run the tests. `qemu` is shorthand for
>  
>  #### In a VM, Using the PPA
>  
> -    $ autopkgtest -U -s -o dep8-mypackage-ppa --setup-commands="sudo add-apt-repository -y -u -s ppa:mylaunchpaduser/bionic-mypackage-fixed-something-1234567" -B mypackage -- qemu /var/lib/adt-images/autopkgtest-bionic-amd64.img
> +    $ autopkgtest -U -s -o dep8-mypackage-ppa --setup-commands="sudo add-apt-repository -y -u -s ppa:mylaunchpaduser/focal-mypackage-fixed-something-1234567" -B mypackage -- qemu /var/lib/adt-images/autopkgtest-focal-amd64.img
>  
> 
> As it is a common issue once you need cross-arch, what do you think about also adding a small
> "how to run in canonistack" section based on and linking to 
> https://wiki.canonical.com/InformationInfrastructure/IS/CanoniStack-BOS01 and https://wiki.ubuntu.com/ProposedMigration#Reproducing_tests_in_the_cloud ?
> 

I think that's a great idea, particularly since that's an area I'm weak
in so would benefit from that myself!

Would you be willing to submit a separate PR with a draft of that?  I'll
be happy to serve as guinea pig and reviewer.  :-)

Bryce

utkarsh2102 commented:
>> As it is a common issue once you need cross-arch, what do you think about also adding a small
 "how to run in canonistack" section based on and linking to
 https://wiki.canonical.com/InformationInfrastructure/IS/CanoniStack-BOS01 and https://wiki.ubuntu.com/ProposedMigration#Reproducing_tests_in_the_cloud ?
>
> I think that's a great idea, particularly since that's an area I'm weak
in so would benefit from that myself!
>
> Would you be willing to submit a separate PR with a draft of that?  I'll
be happy to serve as guinea pig and reviewer.  :-)

I much prefer having a separate document rather than a separate section. The page could then be linked to wherever it seems appropriate. If it makes sense, I can write this up on Monday since I've recently started using canonistack and have been using it frequently. And maybe @cpaelzer could add the missing bits on top of this with his pro-experience 😄

bryceharrington commented:
> > > As it is a common issue once you need cross-arch, what do you think about also adding a small
> > > "how to run in canonistack" section based on and linking to
> > > https://wiki.canonical.com/InformationInfrastructure/IS/CanoniStack-BOS01 and https://wiki.ubuntu.com/ProposedMigration#Reproducing_tests_in_the_cloud ?
> > 
> > 
> > I think that's a great idea, particularly since that's an area I'm weak
> > in so would benefit from that myself!
> > Would you be willing to submit a separate PR with a draft of that?  I'll
> > be happy to serve as guinea pig and reviewer.  :-)
> 
> I much prefer having a separate document rather than a separate section. The page could then be linked to wherever it seems appropriate. If it makes sense, I can write this up on Monday since I've recently started using canonistack and have been using it frequently. And maybe @cpaelzer could add the missing bits on top of this with his pro-experience smile

Yeah if there's enough to say to make it a separate section, that makes sense.  So yep +1.

The balancing act to strike is to let the external pages like the wiki be the formal official reference documentation, and this document be more like the training guidance and advice for someone that may be completely new to the concepts.

utkarsh2102 commented:
Thanks for your work, @bryceharrington. Merging in as discussed! 🎉

cpaelzer commented:
> > > > As it is a common issue once you need cross-arch, what do you think about also adding a small
> > > > "how to run in canonistack" section based on and linking to
> > > > https://wiki.canonical.com/InformationInfrastructure/IS/CanoniStack-BOS01 and https://wiki.ubuntu.com/ProposedMigration#Reproducing_tests_in_the_cloud ?
> > > 
> > > 
> > > I think that's a great idea, particularly since that's an area I'm weak
> > > in so would benefit from that myself!
> > > Would you be willing to submit a separate PR with a draft of that?  I'll
> > > be happy to serve as guinea pig and reviewer.  :-)
> > 
> > 
> > I much prefer having a separate document rather than a separate section. The page could then be linked to wherever it seems appropriate. If it makes sense, I can write this up on Monday since I've recently started using canonistack and have been using it frequently. And maybe @cpaelzer could add the missing bits on top of this with his pro-experience smile
> 
> Yeah if there's enough to say to make it a separate section, that makes sense. So yep +1.
> 
> The balancing act to strike is to let the external pages like the wiki be the formal official reference documentation, and this document be more like the training guidance and advice for someone that may be completely new to the concepts.

@bryce / @utkarsh2102 -  I've given this a first stab at #12
